### PR TITLE
More Chat Service Parity

### PR DESCRIPTION
**Pull request by JackUrb**
## Patch description
**TL:DR;** External chat services were only built to handle one single pass of a task world, and would treat subsequent messages as new instances rather than one continuous flow. This PR restores the idea of returning to a fresh overworld, and also includes the fixes that let us direct from task world to task world. This should increase the parity between OSS and our internal stuff, letting us deploy the same worlds in both places.

### Extended:
FB-Internal chat worlds were using a specific task flow surrounding launching overworlds, onboarding worlds, and tasks worlds, generally splitting into having the overworld and onboarding handled by one thread and then the task later. This behavior was mirrored here in Chat Services, except for the fact that in FB-Intern we allow the overworld/onboarding thread to exit and create a new one when the agent wants to return, but chat services would return someone to the long-running overworld thread that already existed. This was an old, somewhat inefficient model that we've deprecated internally, and so I've tried to update the external code with as minimal changes as possible to get to parity. 

This involved refactoring the `_done_callback` method for a task world (now in `_get_done_callback_for_agents`), the process for launching an overworld (now in `_launch_overworld` rather than hidden within `_on_first_message`), and the process for removing an agent for real (now in `_remove_agent`, not as a part of the task world's `_done_callback`).

I also updated the `TerminalChat` UI to show quick replies 👍 

## Testing steps
I launched the overworld demo task and tried using the onboard data task more than once in a row. I also added an EXIT option to the overworld to ensure that it would dump my information (and future messages would appear as a new user).

### Return to overworld test
```
[ Terminal Chat: ] 
[  port: 1234 ]
[ Current ParlAI commit: 4a1f4e06166dedabf89545c6d7256b9d66395943 ]
[ Current internal commit: 4558a046a633fa1b54869231e45d7dc042dcc938 ]
Connecting to port:  1234
 Me: ParlAI

Bot: Welcome to the overworld for the ParlAI messenger demo. Choose one of the demos from the listed quick replies. 

Options: [echo|onboard data|chat]

 Me: onboard data

Bot: Transferring to onboard data


Bot: Welcome to the onboarding world the onboarding data demo.<br />Enter your name.

 Me: Jack

Bot: <br />Enter your favorite color.

 Me: Blue

Bot: During onboarding, you said your name was Jack and your favorite color was Blue


Bot: Welcome to the overworld for the ParlAI messenger demo. Choose one of the demos from the listed quick replies. 

Options: [echo|onboard data|chat]

 Me: echo

Bot: Transferring to echo


Bot: Welcome to the onboarding world for our echo bot. The next message you send will be echoed. Use [DONE] to finish the chat.

 Me: beep

Bot: beep

 Me: [DONE]
```
### Exit test
```
[ Terminal Chat: ] 
[  port: 1234 ]
Connecting to port:  1234
 Me: ParlAI

Bot: Welcome to the overworld for the ParlAI messenger demo. Choose one of the demos from the listed quick replies. 

Options: [echo|onboard data|chat|EXIT]

E Me: XIT

Bot: Transferring to EXIT


Bot: See you later!

 Me: hello

Bot: Sorry, this conversation bot is password-protected. If you have the password, please send it now.

 Me: ParlAI

Bot: Welcome to the overworld for the ParlAI messenger demo. Choose one of the demos from the listed quick replies. 

Options: [echo|onboard data|chat|EXIT]

 Me: EXIT

Bot: Transferring to EXIT


Bot: See you later!
```

# Discussion
I don't actually know how/if this interfaces with any other projects we are currently working on relying on `ChatService`, in fact I'm not sure how they could've been properly working before (insofar as getting into more than one task world). Happy to do additional testing elsewhere if needed.

### CODE REVIEW

File: parlai/chat_service/services/terminal_chat/client.py [hunk 1]
```diff
@@ -36,7 +36,12 @@ def on_message(ws, message):
     :param message: json with 'text' field to be printed
     """
     incoming_message = json.loads(message)
-    print("\033[0m\nBot: " + incoming_message['text'], "\033[44m\n")
+    print("\033[0m\n")
+    print("Bot: " + incoming_message['text'])
+    quick_replies = incoming_message.get('quick_replies')
+    if quick_replies is not None and len(quick_replies) > 0:
+        print(f"\nOptions: [{'|'.join(quick_replies)}]")
```
Comment (klshuster) [inline parlai/chat_service/services/terminal_chat/client.py]: this is awesome

File: parlai/chat_service/tasks/overworld_demo/worlds.py [hunk 2]
```diff
@@ -252,6 +252,7 @@ class MessengerOverworld(World):
             MessengerOnboardDataTaskWorld,
         ),
         'chat': (MessengerChatOnboardWorld, MessengerChatTaskWorld),
+        'EXIT': (None, None),
```
Comment (klshuster) [inline parlai/chat_service/tasks/overworld_demo/worlds.py]: this means that we'll always need `EXIT` as an option? or is this just specific to this setup

File: parlai/chat_service/core/chat_service_manager.py [hunk 3]
```diff
@@ -614,7 +629,28 @@ def _done_callback(fut):
             for agent in agents:
                 self.after_agent_removed(agent.id)
                 agent_state = self.get_agent_state(agent.id)
-                agent_state.set_active_agent(agent_state.get_overworld_agent())
+                next_task = agent.data.get("next_task")
+                shared_utils.print_and_log(
+                    logging.INFO, "Next task: {}".format(next_task)
+                )
+                if next_task is None:
```
Comment (klshuster) [inline parlai/chat_service/core/chat_service_manager.py]: was this the default behavior before?

File: parlai/chat_service/core/manager_utils.py [hunk 4]
```diff
@@ -345,13 +349,7 @@ def _world_function():
                     agent_state.onboard_data = onboard_data
                 self.manager.add_agent_to_pool(agent_state, world_type)
                 utils.print_and_log(logging.INFO, 'onboarding/overworld complete')
-                time.sleep(5)
 
-                # sleep until agent returns from task world
-                while agent_state.get_active_agent() != overworld_agent:
-                    time.sleep(2)
-                overworld.return_overworld()
```
Comment (klshuster) [inline parlai/chat_service/core/manager_utils.py]: so now `return_overworld` is never called? should we remove these funcs from the demos?

**Comment by klshuster** [comment]
nothing much else currently relies on the `ChatService` framework so i think it's safe to incorporate these changes

File: parlai/chat_service/core/manager_utils.py (hunk 4 above)
Comment (JackUrb) [reply parlai/chat_service/core/manager_utils.py]: Yeah `return_overworld` was a function that we proposed for the initial messenger overworld API but never actually ended up needing. I can remove all callsites and defs

File: parlai/chat_service/core/chat_service_manager.py (hunk 3 above)
Comment (JackUrb) [reply parlai/chat_service/core/chat_service_manager.py]: Default behavior here in external ParlAI was to dump you into a completed overworld, which would just end a chat entirely. This reflects the default in our internal implementation.

File: parlai/chat_service/tasks/overworld_demo/worlds.py (hunk 2 above)
Comment (JackUrb) [reply parlai/chat_service/tasks/overworld_demo/worlds.py]: It's sorta up in the air as to if a service wants to implement an explicit exit. People can always just disconnect
